### DISCUSSION TITLE
[Troubleshooting] NIC Teaming on Cobbler System

### DISCUSSION THREAD

**Discussion by ericchambers1940** [cobbler/cobbler#2751] [Version specific questions]
Hello!

I'm having difficulty configuring NIC teaming on my client system. The client bare metal server pxe boots, installs the OS images, and kickstarts without error. I get to a prompt and run "ip addr". However, no bond0 (name I gave for bonding interface) shows up. The main interface was configured fine and has network connectivity. 

One thing to note is the second network interface shows up but is not connected to the switch right now.

My Environment:
- Oracle Linux 8.3 running the latest available cobbler 3.2.0.
- Dell Poweredge r740 as the PXE client. 
- Cobbler distro uses Oracle Linux 7.8 (vendor ISO).
- I have a custom kickstart (code below).

Here are the commands I ran to setup the system:

```
 1419  2021-08-10 11:03:21 cobbler system add --name=uts0stl23 --profile=ol78-x86_64
 1420  2021-08-10 11:03:22 cobbler system edit --name=uts0stl23 --hostname=uts0stl23.omitted.int --dns-name=uts0stl23.omitted.int
 1421  2021-08-10 11:03:22 cobbler system edit --name=uts0stl23 --interface=em1 --mac=e4:43:4b:94:aa:b4 --interface-type=bond_slave --interface-master=bond0
 1422  2021-08-10 11:03:23 cobbler system edit --name=uts0stl23 --interface=em3 --mac=e4:43:4b:94:aa:b6 --interface-type=bond_slave --interface-master=bond0
 1423  2021-08-10 11:03:25 cobbler system edit --name=uts0stl23 --interface=bond0 --interface-type=bond --bonding-opts="miimon=100 mode=4 xmit_hash_policy=1" --ip-address=10.158.129.191 --netmask=255.255.252.0 --gateway=10.158.129.191 --static=1
 1424  2021-08-10 11:03:33 cobbler system report
 1427  2021-08-10 11:08:50 cobbler sync
```

The sync ran fine and generated my new DHCP reservation:
```
group {
    host generic1 {
        hardware ethernet e4:43:4b:94:aa:b4;
        fixed-address 10.158.129.191;
        option host-name "uts0stl23.omitted.int";
        option subnet-mask 255.255.252.0;
        option routers 10.158.129.191;
        next-server 10.158.128.227;
    }
}
```

Here is my kickstart file with organizational references omitted:
```
[root@uts0stl21 templates]# cat standard.ks
auth  --useshadow  --enablemd5
bootloader --location=mbr
clearpart --all --initlabel
text
firewall --disabled
firstboot --disable
keyboard us
lang en_US
selinux --disabled
url --url=$tree
reboot
rootpw --iscrypted $default_password_crypted
selinux --disabled
skipx
timezone --ntpservers=_omitted_ America/Chicago
install
zerombr
clearpart --all --drives=sda,sdb --initlabel
bootloader --location=mbr --driveorder=sda,sdb
part raid.01 --asprimary --fstype="raid" --ondisk=sda       --size=512
part raid.02             --fstype="raid" --grow       --ondisk=sdb --size=1
part raid.03 --asprimary --fstype="raid" --ondisk=sda              --size=512
part raid.04             --fstype="raid" --grow       --ondisk=sdb --size=1
raid /boot --device=md0 --fstype="xfs" --level=1 raid.01 raid.03
raid pv.01 --fstype "physical volume (LVM)" --level=RAID1 --device=md1 raid.02 raid.04
volgroup rootvg --pesize=32768 pv.01
logvol /                --fstype xfs --name=root         --vgname=rootvg --size=4096
logvol swap             --fstype swap --name=swap         --vgname=rootvg --size=4096
logvol /var             --fstype xfs --name=var          --vgname=rootvg --size=4096
logvol /var/crash       --fstype xfs --name=crash        --vgname=rootvg --size=16384
logvol /tmp             --fstype xfs --name=tmp          --vgname=rootvg --size=2048
logvol /uts_staging     --fstype xfs --name=uts_staging  --vgname=rootvg --size=4096
logvol /nmon_data       --fstype xfs --name=nmon_data    --vgname=rootvg --size=1024
logvol /image           --fstype xfs --name=image        --vgname=rootvg --size=4096
logvol /altroot         --fstype xfs --name=altroot      --vgname=rootvg --size=4096    --fsoptions="noauto"
logvol /altvar          --fstype xfs --name=altvar       --vgname=rootvg --size=4096    --fsoptions="noauto"
%pre
$SNIPPET('log_ks_pre')
$SNIPPET('autoinstall_start')
$SNIPPET('pre_install_network_config')
$SNIPPET('pre_anamon')
%end
%packages
    autofs
    bind-utils
    cups
    cyrus-sasl-gssapi
    dmidecode
    emacs
    finger
    fuse
    fuse-libs
    gawk
    git
    iotop
    ksh
    libXtst
    MySQL-python
    nfs-utils
    ntp
    openldap
    openldap-clients
    openldap-devel
    openssh-clients
    openssl
    openssl-devel
    perf
    perl
    python-psycopg2
    redhat-lsb
    redhat-lsb-core
    sg3_utils
    subversion
    sysfsutils
    telnet
    unzip
    vim-enhanced
    wget
    xorg-x11-fonts*
    xorg-x11-font-utils
    xorg-x11-xauth
%end
%post
systemctl enable ntpd
service ntpd restart
%end
%post
curl -Sks "https://artifacts.blah.int:443/artifactory/archive-external-release/com/blah/operations/linux/spacewalk/bootstrap-mc_spacewalk_210.s" | /bin/bash
%end
```

Other things I checked:
-I looked in the dir /etc/sysconfig/network-scripts/* to see if bond0 exists as a file there. It's missing.

Please let me know what you think,
Eric

**Comment by SchoolGuy** [comment]
So I will repeat myself in regard to what I said in the Gitter already. This part of the code is not tested. I will create a testcase for it and see what I can do. Will take a few days though because I am working on multiple things already.

@nodeg We should discuss tomorrow what machine we can snatch for this from the pool and attach it to your current test-instance of Cobbler. I believe the `master` code might fail us miserably since we havn't performed any checks. I mean we talked about the NetworkInterface tests but I believe it is time we cover that part finally also with automated tests.

@heroin-moose Could you add that at least as an empty TODO script to the setup that you are developing, I believe we can take the script from above already for reproduction.

**Reply by ericchambers1940** [reply]
Thank you @SchoolGuy, as always. I’ll keep troubleshooting on my end!

**Reply by heroin-moose** [reply]
Yeah, sure.

**Reply by ericchambers1940** [reply]
I think I found a potential cause if this issue:

In my kickstart template, I was utilizing $SNIPPET('pre_install_network_config') to generate the network configuration it seems.

Looking at that snippet file on my Cobbler test server, it seems most if not all bonding related logic is commented out. I think this snippet utilizes the bonding variables shown in "cobbler system dumpvar" (which are set using cobbler system edit). 

Not sure why they are commented out, but I wonder if that commented code is what's needed for bonding to be setup.

It would be nice to be able to setup bonding via the built in snippet. In the meantime, I wrote my own snippet that simply generates the bonding master and slave interfaces. I tested this and it works well. Bond0 shows up fine using 'ip addr' and /proc/net/bonding/bond0 shows the correct bonding options are running. This lowers the priority of this issue for now.

Here's the snippet in case anyone else needs it:

```
%post --log=/root/ks_nic_bonding.log
Load the bonding kernel module
modprobe --first-time bonding

cat << EOF1 >/etc/sysconfig/network-scripts/ifcfg-bond0
DEVICE=bond0
BOOTPROTO=static
ONBOOT=yes
IPADDR=$ip
NETMASK=$mask
MTU=1500
BONDING_OPTS=$bondingopts
DNS1=$dns1
GATEWAY=$gateway
EOF1

cat << EOF2 >>/etc/sysconfig/network-scripts/ifcfg-em1
DEVICE=em1
MASTER=bond0
ONBOOT=yes
BOOTPROTO=static
SLAVE=yes
EOF2

cat << EOF3 >>/etc/sysconfig/network-scripts/ifcfg-em3
DEVICE=em3
MASTER=bond0
ONBOOT=yes
BOOTPROTO=static
SLAVE=yes
EOF3

echo 'alias bond0 bonding' >> /etc/modprobe.conf

systemctl restart network
%end
```


I used the "--autoinstall-meta=" argument for my test system to define the variables shown above:
- $bondingopts was set to: "miimon=100 mode=4 xmit_hash_policy=layer3+4"
- $ip, $mask, $dns1, $gateway are self explanatory

**Reply by SchoolGuy** [reply]
@ericchambers1940 If we are able to generalize your snippet we could replace it with the commented out code if you want. Those snippets are currently not getting the love they need, so I wouldn't mind if yours is an improvement to the old one.

**Reply by ericchambers1940** [reply]
@SchoolGuy id be happy to generalize the code and contribute it to the project. However, I cannot do it from work (where there’s plenty of testable bare metal hardware) since essentially any code I write there belongs to them. Therefore, I’ll be limited on testing. 

I could potentially spin up a VM on my ESXi server at home and set it to pxe boot. Do you consider that adequate testing?

**Reply by SchoolGuy** [reply]
@ericchambers1940 Yes that is plenty of testing.

**Reply by ericchambers1940** [reply]
Hey @SchoolGuy, it’s gonna be a little bit before I can get around to this. I’m scrambling to wrap up configuring cobbler for our environment and have two weeks left. I’ll look into this then.
